nuxeo-features/nuxeo-thumbnail/src/main/java/org/nuxeo/ecm/platform/thumbnail/action/RecomputeThumbnailsAction.java

```diff
@@ -0,0 +1,153 @@
+            xpath = command.getParam(PARAM_XPATH);
+        }
+
+        @Override
+        protected void compute(CoreSession session, List<String> ids, Map<String, Serializable> properties) {
+
+            DocumentModelList docs = loadDocuments(session, ids);
+
+            for (DocumentModel doc : docs) {
+                processDoc(session, doc);
+            }
+        }
+
+        protected void processDoc(CoreSession session, DocumentModel doc) {
+            Blob thumbnailBlob = getManagedThumbnail(doc);
+            if (thumbnailBlob == null) {
+                ThumbnailAdapter thumbnailAdapter = doc.getAdapter(ThumbnailAdapter.class);
+                if (thumbnailAdapter == null) {
+                    return;
+                }
+                thumbnailBlob = thumbnailAdapter.computeThumbnail(session);
+            }
+            if (thumbnailBlob != null) {
+                if (!doc.hasFacet(ThumbnailConstants.THUMBNAIL_FACET)) {
+                    doc.addFacet(ThumbnailConstants.THUMBNAIL_FACET);
+                }
+                doc.setPropertyValue(ThumbnailConstants.THUMBNAIL_PROPERTY_NAME, (Serializable) thumbnailBlob);
+            } else {
+                if (doc.hasFacet(ThumbnailConstants.THUMBNAIL_FACET)) {
+                    doc.setPropertyValue(ThumbnailConstants.THUMBNAIL_PROPERTY_NAME, null);
+                    doc.removeFacet(ThumbnailConstants.THUMBNAIL_FACET);
+                }
+            }
+            if (doc.isDirty()) {
+                doc.putContextData(VersioningService.VERSIONING_OPTION, VersioningOption.NONE);
+                doc.putContextData(VersioningService.DISABLE_AUTO_CHECKOUT, Boolean.TRUE);
+                doc.putContextData(DublinCoreListener.DISABLE_DUBLINCORE_LISTENER, Boolean.TRUE);
+                doc.putContextData(NotificationConstants.DISABLE_NOTIFICATION_SERVICE, Boolean.TRUE);
+                doc.putContextData("disableAuditLogger", Boolean.TRUE);
+                if (doc.isVersion()) {
+                    doc.putContextData(ALLOW_VERSION_WRITE, Boolean.TRUE);
+                }
+                doc.putContextData(THUMBNAIL_UPDATED, true);
+                session.saveDocument(doc);
+            }
+        }
+
+        protected Blob getManagedThumbnail(DocumentModel doc) {
+            BlobHolder bh = doc.getAdapter(BlobHolder.class);
+            if (bh == null) {
+                return null;
+            }
+            Blob blob = bh.getBlob();
+            if (blob == null) {
+                return null;
+            }
+            BlobManager blobManager = Framework.getService(BlobManager.class);
+            try (InputStream is = blobManager.getThumbnail(blob)) {
+                if (is == null) {
+                    return null;
+                }
+                return Blobs.createBlob(is);
+            } catch (IOException e) {
+                throw new NuxeoException("Failed to get managed blob thumbnail", e);
+            }
+        }
+    }
+}
```

Comment:
add break line at the end of the class, (usually the missing line at the of file is reported by github) i don't why it doesn't this time